nullaway/src/test/java/com/uber/nullaway/NullAwayContractsBooleanTests.java

```diff
@@ -0,0 +1,436 @@
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void nullCheckIsFalseIsNotNullableReversed() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test(@Nullable Object o1) {",
+            "    Validation.checkFalse(null == o1);",
+            "    return o1.toString();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void nullCheckIsTrueIsNull() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test(@Nullable Object o1) {",
+            "    Validation.checkTrue(o1 == null);",
+            "    // BUG: Diagnostic contains: dereferenced expression",
+            "    return o1.toString();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void nullCheckIsTrueIsNullReversed() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test(@Nullable Object o1) {",
+            "    Validation.checkTrue(null == o1);",
+            "    // BUG: Diagnostic contains: dereferenced expression",
+            "    return o1.toString();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void nonNullCheckIsFalseIsNull() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test(@Nullable Object o1) {",
+            "    Validation.checkFalse(o1 != null);",
+            "    // BUG: Diagnostic contains: dereferenced expression",
+            "    return o1.toString();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void nonNullCheckIsFalseIsNullReversed() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test(@Nullable Object o1) {",
+            "    Validation.checkFalse(null != o1);",
+            "    // BUG: Diagnostic contains: dereferenced expression",
+            "    return o1.toString();",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void compositeNullCheckAndStringEquality() {
+    helper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import javax.annotation.Nullable;",
+            "class Test {",
+            "  String test(@Nullable Object o1) {",
+            "    Validation.checkTrue(o1 != null && o1.toString().equals(\"\"));",
```

Comment:
This is pretty damn cool! Do we have the check for the case where we have this same code but `||` instead of `&&`? Are we correctly detecting that `o1` may still be nullable after?

Reply:
Good idea, I've added additional cases to this test